Hw01: Приложение по проведению тестирования студентов - только вывод вопросов и вариантов ответа (если имеются)

SmershRoot opened the pull request:
**Реализация ДЗ1.**
 - Источник данных (csv) дополнено до 5-и вопросов.
 - Настройка компонентов выполнена
 - Тестовый класс реализован для CsvQuestionDao 4 метода для тестирования получения данных findAll()
 - Checkstyle - пройден

**Задание со звездочкой**
Приложение должно корректно запускаться с помощью "java -jar" - выполнено. Но при сборке пакета Warning: "maven-shade-plugin обнаружил, что некоторые файлы классов присутствует в двух или более JAR-файлах. Работать -работает. Но позже придется все равно разобраться детальнее и убрать этот warning"

**Запрашиваемый критерий оценки\приема работы**
Цель стать сильно лучше чем есть сейчас. Буду благодарен если помимо ошибок Вы укажите на то где я недоработал и где можно сделать лучше. Спасибо.

SmershRoot committed:
hw01: Начальная настройка контекста шаблона

SmershRoot committed:
hw01: Чтение данных из csv-файла

SmershRoot committed:
hw01: Добавил вопросы до 5-и. Добавил вывод их в консоль

SmershRoot committed:
hw01: Настроил maven-shade-plugin

Warning идет: maven-shade-plugin обнаружил, что некоторые файлы классов
присутствует в двух или более JAR-файлах. Работать -работает. Но тут надо разбираться и убирать

SmershRoot committed:
hw01: Добавил тесты на findAll()

SmershRoot committed:
hw01: test: Добавил тестовый метод проверки тестового вопроса и его ответов

stvort commented on hw01/src/main/java/ru/otus/hw/dao/CsvQuestionDao.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Override
+    public List<Question> findAll() {
+        var inputStream = getClass().getResourceAsStream(fileNameProvider.getTestFileName());
```
Ресурс за собой лучше закрыть

SmershRoot replied:
Сделал как:
try (var inputStream = getClass().getResourceAsStream(fileNameProvider.getTestFileName())) {

и добавил обработку исключений в перечне catch 
 } catch (QuestionReadException e) {
            throw new QuestionReadException(e.getMessage());
        } catch (IOException e) {
            throw new QuestionReadException("Ошибка взаимодействия с источником данных", e);
        } catch (Exception e) {
            throw new QuestionReadException("Ошибка чтения данных", e);
        }

stvort commented on hw01/src/main/java/ru/otus/hw/service/TestServiceImpl.java:
```diff
@@ -0,0 +1,28 @@
+        ioService.printFormattedLine("Please answer the questions below%n");
+
+        var questions = questionDao.findAll();
+        questions.forEach(question -> {
+            ioService.printFormattedLine("%d. %s", questions.indexOf(question) + 1, question.text());
+            var answers = question.answers();
+            answers.forEach(answer -> {
+                ioService.printFormattedLine("- %d. %s", answers.indexOf(answer) + 1, answer.text());
+            });
```
Это лучше вынести в метод

SmershRoot replied:
Вынес в методы вывода вопросов и ответов.

stvort commented on hw01/src/main/java/ru/otus/hw/service/TestServiceImpl.java:
```diff
@@ -0,0 +1,28 @@
+
+        var questions = questionDao.findAll();
+        questions.forEach(question -> {
+            ioService.printFormattedLine("%d. %s", questions.indexOf(question) + 1, question.text());
```
Ну и вообще не обязательно использовать лямбды там, где отлично подходит `for` со счетчиком

SmershRoot replied:
исправил

stvort commented on hw01/src/test/java/ru/otus/hw/dao/CsvQuestionDaoTest.java:
```diff
@@ -0,0 +1,92 @@
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class CsvQuestionDaoTest {
```
> Крайне желательно написать юнит-тест какого-нибудь **сервиса**

Дао не сервис, а тест, в которым задействована файловая система не может считаться юнит-тестом

SmershRoot replied:
Спасибо. Протестировать findAll() на различное поведение первое что пришло в голову и показалось интересным, и упустил то что сделал не то что требовалось по задаче.
ДАО - не стал убирать тест, но пометил его описанием, что он не unit.

StreamsIOServiceTest - более-менее стандартный и простой.
TestServiceImplTest - проверка на то что вопросы и варианты ответов полученные из источника поступают на вывод без модификаций

stvort commented on hw01/src/test/java/ru/otus/hw/dao/CsvQuestionDaoTest.java:
```diff
@@ -0,0 +1,92 @@
+    private static final List<String> TEST_ANSWER_CORRECT = List.of("Vasily Oleynik", "Irina Akhmadullina");
+    private static final List<String> TEST_ANSWER_NOT_CORRECT = List.of("Maxim Orlovsky", "Warren Buffett");
+
+    @Mock
+    TestFileNameProvider fileNameProvider;
+
+    @InjectMocks
+    CsvQuestionDao csvQuestionDao;
```
`private`

SmershRoot replied:
исправил

SmershRoot committed:
hw01: fix: CsvQuestionDao. inputStream clouse.  Переделал исключения

SmershRoot committed:
hw01: test: Вынес с методы вывод вопросов и списка ответов к ним

SmershRoot committed:
hw01: fix: удалил лишнюю строку

SmershRoot committed:
hw01: fix: добавил в ДАО-тест комментарий и private

Добавил комментарий про то что тест где задействовано файловое хранилище не является юнит-тестом. Тут скорее для себя, чтобы лучше запомнил

SmershRoot committed:
hw01: тест StreamsIOService

SmershRoot committed:
hw01: Тест TestServiceImplTest

Очень хотелось проверить содержимое вывода. И написал вот такой тест.

SmershRoot committed:
hw01: рефакторинг теста

stvort commented on hw01/src/main/java/ru/otus/hw/dao/CsvQuestionDao.java:
```diff
@@ -0,0 +1,42 @@
+    public List<Question> findAll() {
+        try (var inputStream = getClass().getResourceAsStream(fileNameProvider.getTestFileName())) {
+            if (Objects.isNull(inputStream)) {
+                throw new QuestionReadException("Источник получения данных не доступен");
```
> Весь ввод-вывод осуществляется на английском языке

Странно, что исключения на русском при этом

SmershRoot replied:
1.Тексты ошибок на английском. Тесты оставил на русском. Думаю они не ввод-вывод (но на всякий случай подготовил коммит с их переводом).

stvort commented on hw01/src/main/java/ru/otus/hw/dao/CsvQuestionDao.java:
```diff
@@ -0,0 +1,42 @@
+                    .build()
+                    .parse();
+            return questionDtos.stream().map(QuestionDto::toDomainObject).toList();
+        } catch (QuestionReadException e) {
+            throw new QuestionReadException(e.getMessage());
```
Зачем преобразовывать `QuestionReadException` в самого себя? Можно кинуть оригинал или вовсе не ловить

SmershRoot replied:
Сделал так что кидаю оригинал: throw e;

stvort commented on hw01/src/test/java/ru/otus/hw/service/StreamsIOServiceTest.java:
```diff
@@ -0,0 +1,46 @@
+    @DisplayName(value = "Тест метода ioService.printFormattedLine должен выводить сообщение в консоль с параметрами")
+    void printFormattedLine() {
+        String testString = "Message for test";
+        String string_arg = "string_arg";
```
По соглашению, в Java переменные именуются в `lowerCamelCase` и никак иначе)

SmershRoot replied:
3. Поправил именование переменных.

stvort commented on hw01/src/test/java/ru/otus/hw/service/StreamsIOServiceTest.java:
```diff
@@ -0,0 +1,46 @@
+
+    /**
+     * Тест метода ioService.printFormattedLine должен выводить сообщение в консоль с параметрами
+     * <p>Посчитал что нормальное поведение если текст перед вызовом printf может быть в будущем модифицирован,
```
Если будет модифицирован, то мы его потом и проверим т.к. тесты начнут падать. Ну и в этом случае результат ожидаемый не сойдется. Сейчас, если нам вообще нужна проверка форматированной строки, то ее и нужно проверять

SmershRoot replied:
4. StreamsIOServiceTest - проверяю теперь формат первого атрибута printf в том виде как он описан в коде

Извините, не удержался и дополнил тест тем что теперь текст генерируется случайным образом и выполняется тест 5 раз. По ощущениям такое построение теста более правильное.

stvort commented on hw01/src/test/java/ru/otus/hw/service/TestServiceImplTest.java:
```diff
@@ -0,0 +1,67 @@
+
+    @BeforeEach
+    public void setUp() {
+        PrintStream printStream = new PrintStream(outputStreamCaptor);
+        IOService ioService = new StreamsIOService(printStream);
```
Нет. Мы пишем юнит-тест. Тут все зависимости являются моками, не моком будет только тестируемый класс

SmershRoot replied:
5. TestServiceImplTest - переписал под то что не Mock только один сервис который и тестируем.

stvort commented on hw01/src/test/java/ru/otus/hw/service/TestServiceImplTest.java:
```diff
@@ -0,0 +1,67 @@
+    public void setUp() {
+        PrintStream printStream = new PrintStream(outputStreamCaptor);
+        IOService ioService = new StreamsIOService(printStream);
+        questionDao = mock(QuestionDao.class);
```
Давайте все моки делать во всех тестах единообразно

SmershRoot replied:
Исправил

SmershRoot committed:
hw01: hix: Тексты ошибок на английском. Тесты оставил на русском. Думаю они не ввод-вывод

SmershRoot committed:
hw01: fix: printf проверка точного совпадения. Генерация тестовых данных

SmershRoot committed:
hw01: fix: TestServiceImplTest только с одним проверяемым не Mock

От идеи проверить вывод итоговый на содержимое полученных из источника данных не отказался. Но надо поэкспериментировать на возможность сделать это с Mock-ами не тестируемого сервиса.
Может и не получиться и это не уложиться в правильный unit-test и будет тест другого типа.

stvort commented on hw01/src/test/java/ru/otus/hw/service/StreamsIOServiceTest.java:
```diff
@@ -29,18 +31,27 @@ void printLine() {
-        int int_arg = 1;
-        ioService.printFormattedLine(testString, string_arg, int_arg);
-        verify(printStream, times(1)).printf(anyString(), eq(string_arg), eq(int_arg));
+        String testString = generateRandomString(100);
+        String stringArg = generateRandomString(10);
+        int intArg = new Random().nextInt();
```
Юнит-тест должен быть возобновляемым, чего не гарантировано при случайной генерации аргументов
